exercise/cache/points.py

```diff
@@ -133,21 +210,103 @@ def r_augment(children: List[Dict[str, Any]]) -> None:
+                that feedback is hidden appropriately.
+                """
+                rule = exercise.active_submission_feedback_reveal_rule
+                state = ExerciseRevealState(entry)
+                is_revealed = rule.is_revealed(state)
+                reveal_time = rule.get_reveal_time(state)
+
+                entry.update({
+                    'best_submission': entry['best_submission'] if is_revealed else last_submission.id,
+                    'points': entry['points'] if is_revealed else 0,
+                    'formatted_points': format_points(entry['points'], is_revealed, False),
+                    'passed': entry['passed'] if is_revealed else False,
+                    'feedback_revealed': is_revealed,
+                    'feedback_reveal_time': reveal_time,
+                })
+
+                for submission in entry['submissions']:
+                    submission.update({
+                        'points': submission['points'] if is_revealed else 0,
+                        'formatted_points': format_points(submission['points'], is_revealed, False),
+                        'passed': submission['passed'] if is_revealed else False,
+                        'feedback_revealed': is_revealed,
+                        'feedback_reveal_time': reveal_time,
+                    })
+
+                # If the reveal rule depends on time, update the cache's
+                # invalidation time.
+                if (
+                    reveal_time is not None
+                    and reveal_time > timezone.now()
+                    and (
+                        data['invalidate_time'] is None
+                        or reveal_time < data['invalidate_time']
+                    )
+                ):
+                    data['invalidate_time'] = reveal_time
+
+            # Augment submission data.
+            for submission in submissions:
+                # The submissions are ordered by exercise. Check here if the
+                # exercise has changed.
+                if exercise is None or submission.exercise.id != exercise.id:
+                    if exercise is not None and not is_staff:
+                        # Check the reveal rule of the last exercise now that
+                        # all of its submissions have been iterated.
+                        check_reveal_rule()
+                    final_submission = None
+                    last_submission = None
+                    # These variables stay constant throughout the exercise.
+                    exercise = submission.exercise
+                    try:
+                        tree = self._by_idx(modules, exercise_index[submission.exercise.id])
+                    except KeyError:
+                        self.dirty = True
+                        continue
+                    entry = tree[-1]
+                    if submission.exercise.grading_mode == BaseExercise.GRADING_MODE.BEST:
```

Comment:
`submission.exercise` could be just `exercise` here.

Reply:
Fixed